### PR TITLE
FOSRestController is now a trait and not longer a class extending from the core Controller class

### PR DESCRIPTION
**Pull request by lsmith77**

### CODE REVIEW

**Comment by GromNaN** [comment]
:+1: As PHP 5.4 is required.

**Comment by Ener-Getick** [comment]
I'm wondering if we wouldn't be better to suffix its name with ``Trait`` and to remove ``FOS``: ``RestControllerTrait``.

**Comment by lsmith77** [comment]
yeah .. was pondering this as well

**Comment by Ener-Getick** [comment]
And IMO we should create a last release for the 1.* branch with bc layers to help the migration to FOSRestBundle 2.0.

**Comment by lsmith77** [comment]
not sure if its worth it .. I mean we have tons of open bugs/enhancement tickets .. at this point moving to 2.0 should not take all too long .. what does make sense is to maybe invest some more time into a migration guide

File: UPGRADING.md [hunk 1]
```diff
@@ -16,6 +16,7 @@ This document will be updated to list important BC breaks and behavioral changes
  * Removed ``callback_filter`` configuration option for the jsonp_handler
  * ``exception_wrapper_handler`` is now the name of a service and not the name of a class
  * removed all ``.class`` parameters, instead overwriting services via explicit Bundle configuration is preferred
+ * ControllerTrait is now a trait and not longer a class extending from the core Controller class
```
Comment (Ener-Getick) [inline UPGRADING.md]: This is not clear, imo include the new and the old name would be clearer.

File: Controller/ControllerTrait.php [hunk 2]
```diff
@@ -12,17 +12,23 @@
...
 
 /**
- * Base Controller for Controllers using the View functionality of FOSRestBundle.
+ * Trait for Controllers using the View functionality of FOSRestBundle.
  *
  * @author Benjamin Eberlei <kontakt@beberlei.de>
+ * @author Lukas Kahwe Smith <smith@pooteeweet.org>
  */
-abstract class FOSRestController extends Controller
+trait ControllerTrait
```
Comment (Ener-Getick) [inline Controller/ControllerTrait.php]: I think that this trait must extend the ``ContainerAwareTrait`` as it can't work without it.

Comment (Ener-Getick) [inline Controller/ControllerTrait.php]: Another solution would be to add an abstract method ``get``.

Comment (lsmith77) [inline Controller/ControllerTrait.php]: well the trait falls back to using the magic ``get()`` method .. but it can work without it .. 
I am not quite sure how to best approach this .. maybe provide two traits .. ?

Comment (Ener-Getick) [inline Controller/ControllerTrait.php]: Oops look too fast... 
If there are two traits, the problem will be the same (moreover ``get`` belongs to ``Controller``) ... 
You can maybe check if the ``get`` method exists on ``$this`` or if ``$this`` is an instance of ``Controller``.

File: Controller/ControllerTrait.php [hunk 3]
```diff
@@ -83,6 +89,30 @@ protected function routeRedirectView($route, array $parameters = [], $statusCode
...
+
+        return $this->viewhandler;
+    }
+
+    /**
+     * Get the ViewHandler.
+     *
+     * @@param View $view
+     */
+    protected function setViewHandler(ViewHandlerInterface $viewhandler)
```
Comment (Ener-Getick) [inline Controller/ControllerTrait.php]: Why did you create this function?

Comment (lsmith77) [inline Controller/ControllerTrait.php]: because i wanted this trait also to be useful when not using the ContainerAwareTrait. will send a PR in a few mins that splits that use case into two separate traits

Comment (Ener-Getick) [inline Controller/ControllerTrait.php]: Extending ``getViewHandler`` is already possible and I'm not sure that the setter will be used a lot.

Comment (lsmith77) [inline Controller/ControllerTrait.php]: its useful if you define controllers as services

File: Controller/ControllerTrait.php (hunk 2 above)
Comment (lsmith77) [inline Controller/ControllerTrait.php]: i fixed it by using ``$this->container->get()``

Comment (Ener-Getick) [inline Controller/ControllerTrait.php]: :+1:

File: Controller/ControllerTrait.php [hunk 4]
```diff
@@ -83,6 +89,30 @@ protected function routeRedirectView($route, array $parameters = [], $statusCode
...
+            throw new \RuntimeException('A "ViewHandlerInterface" instance must be set when usign the FOSRestBundle "ControllerTrait".');
+        }
+
+        return $this->viewhandler;
+    }
+
+    /**
+     * Get the ViewHandler.
+     *
+     * @@param View $view
```
Comment (lsmith77) [inline Controller/ControllerTrait.php]: there is a typo here that I need to fix

File: Resources/doc/2-the-view-layer.rst [hunk 5]
```diff
@@ -88,10 +94,12 @@ or it is possible to use lazy-loading:
 
     <?php
 
-    use FOS\RestBundle\Controller\FOSRestController;
+    use Symfony\Bundle\FrameworkBundle\Controller\Controller;
 
-    class UsersController extends FOSRestController
+    class UsersController extends Controller
     {
+        use ContainerAwareControllerTrait;
```
Comment (lsmith77) [inline Resources/doc/2-the-view-layer.rst]: we should actually have at least one test case using this trait

File: Controller/ControllerTrait.php (hunk 2 above)
Comment (Ener-Getick) [inline Controller/ControllerTrait.php]: Finally, I'm wondering if it's not too complicated to explain if there are traits... 
Can't we just check if ``$this`` is an instance of ``ContainerAwareInterface`` or if ``$this->viewHandler`` is defined and otherwise throw an exception?

File: Resources/doc/2-the-view-layer.rst (hunk 5 above)
Comment (lsmith77) [inline Resources/doc/2-the-view-layer.rst]: ```Symfony\Component\DependencyInjection\ContainerAware and FOS\RestBundle\Controller\ContainerAwareControllerTrait define the same property ($container) in the composition of FOS\RestBundle\Tests\Fixtures\Controller\AnnotatedNonPluralizedArticleController. This might be incompatible, to improve maintainability consider using accessor methods in traits instead. Class was composed```

Comment (Ener-Getick) [inline Resources/doc/2-the-view-layer.rst]: You should not use the ``ContainerAwareTrait`` apparently as it is already used by ``Controller``.

**Comment by lsmith77** [comment]
To fix the tests one needs to do the following, which sucks big time as it basically means the trait isn't a logical unit anymore but requires other things that I cannot express with code :-/

```
diff --git a/Controller/ContainerAwareControllerTrait.php b/Controller/ContainerAwareControllerTrait.php
index b03ecd2..d677c1d 100644
--- a/Controller/ContainerAwareControllerTrait.php
+++ b/Controller/ContainerAwareControllerTrait.php
@@ -12,7 +12,6 @@
 namespace FOS\RestBundle\Controller;
 
 use FOS\RestBundle\View\ViewHandlerInterface;
-use Symfony\Component\DependencyInjection\ContainerAwareTrait;
 
 /**
  * Trait for Controllers using the View functionality of FOSRestBundle.
@@ -22,7 +21,6 @@ use Symfony\Component\DependencyInjection\ContainerAwareTrait;
 trait ContainerAwareControllerTrait
 {
     use ControllerTrait;
-    use ContainerAwareTrait;
 
     /**
      * Get the ViewHandler.
```

**Comment by lsmith77** [comment]
Note even if we would change the core ``ContainerAware`` class to use the ``ContainerAwareTrait`` we would still run into the same problem as can be seen here https://3v4l.org/l1qkE

**Comment by lsmith77** [comment]
ok changing to use ``getContainer()`` and defining such an abstract method within ``ContrainerAwareControllerTrait`` along with the following change fixes the issue:
```
diff --git a/ContainerAware.php b/ContainerAware.php
index af977fe..080f8cb 100644
--- a/ContainerAware.php
+++ b/ContainerAware.php
@@ -18,18 +18,5 @@ namespace Symfony\Component\DependencyInjection;
  */
 abstract class ContainerAware implements ContainerAwareInterface
 {
-    /**
-     * @var ContainerInterface
-     */
-    protected $container;
-
-    /**
-     * Sets the Container associated with this Controller.
-     *
-     * @param ContainerInterface $container A ContainerInterface instance
-     */
-    public function setContainer(ContainerInterface $container = null)
-    {
-        $this->container = $container;
-    }
+    use ContainerAwareTrait;
 }
diff --git a/ContainerAwareTrait.php b/ContainerAwareTrait.php
index 57280aa..caf222a 100644
--- a/ContainerAwareTrait.php
+++ b/ContainerAwareTrait.php
@@ -32,4 +32,14 @@ trait ContainerAwareTrait
     {
         $this->container = $container;
     }
+
+    /**
+     * gets the Container associated with this Controller.
+     *
+     * @return ContainerInterface
+     */
+    public function getContainer()
+    {
+        return $this->container;
+    }
 }
```

tl;dr: trait properties should always come with a getter method

**Comment by Ener-Getick** [comment]
What about doing something like https://github.com/FriendsOfSymfony/FOSRestBundle/compare/master...Ener-Getick:CONTROLLER?expand=1 ?

**Comment by lsmith77** [comment]
@Ener-Getick at the end of the day it still relies on a property that is no where defined in the trait or the contract defined by the trait. will see what happens with the ticket I opened on symfony core.

File: Controller/ContainerAwareControllerTrait.php [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+use FOS\RestBundle\View\ViewHandlerInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Trait for Controllers using the View functionality of FOSRestBundle.
+ *
+ * @author Lukas Kahwe Smith <smith@pooteeweet.org>
+ */
+trait ContainerAwareControllerTrait
```
Comment (lsmith77) [inline Controller/ContainerAwareControllerTrait.php]: refactor this to be a base class

**Comment by lsmith77** [comment]
ok .. keeping FOSRestController and just introducing ControllerTrait .. ready for final review.

File: Controller/ControllerTrait.php [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+        return $this->viewhandler;
+    }
+
+    /**
+     * Get the ViewHandler.
+     *
+     * @param ViewHandlerInterface $viewhandler
+     */
+    protected function setViewHandler(ViewHandlerInterface $viewhandler)
```
Comment (Ener-Getick) [inline Controller/ControllerTrait.php]: IMO this should be public to allow setter injection

**Comment by Ener-Getick** [comment]
Except my last comment, sounds good to me :+1:.

File: Controller/ControllerTrait.php (hunk 7 above)
Comment (lsmith77) [inline Controller/ControllerTrait.php]: agreed

**Comment by lsmith77** [comment]
@Ener-Getick fixed
